abcpy/ProbabilisticModel.py

```diff
@@ -0,0 +1,719 @@
+from abc import ABCMeta, abstractmethod
+import numpy as np
+
+from scipy.stats import multivariate_normal, norm
+from scipy.special import gamma
+
+#TODO both ricker and lorenz could support that you give some in a combined model, and some not, we should implement that, but not a priority
+
+#TODO docstrings
+
+#NOTE do we maybe want to average over a couple of samples during initialization, rather than taking a single value? this is an issue for example for StudentT!
+
+class ProbabilisticModel(metaclass = ABCMeta):
+    def __init__(self, parameters):
+        self.parents = parameters #all parents of the node
+        self.updated = False #whether the node has been touched
+
+    #NOTE the check_fixed assumes that the parameters are exactly of the lenght needed for that distribution only ->needs to be considered in case we want to be able to "fix on whole graph"
+    def fix_parameters(self, parameters=None, rng=np.random.RandomState()):
+        if(not(parameters)):
+            self.value = self.sample_from_distribution(1, rng)
+            if(isinstance(self.value[0],np.ndarray)):
+                self.value = self.value[0]
+        else:
+            self.check_parameters_fixed(parameters)
+            index=0
+            i=0
+            while(i<len(parameters)):
+                while(not(isinstance(self.parents[index],ProbabilisticModel)) or self.parents[index].updated):
+                    index+=1
+                if(not(parameters[i])):
+                    self.parents[index].fix_parameters(rng=rng)
+                    self.parents[index].updated = True
+                    index+=1
+                    i+=1
+                else:
+                    for j in range(len(self.parents[index].value)):
+                        self.parents[index].value[j] = parameters[i]
+                        i+=1
+                    self.parents[index].updated = True
+                    index+=1
+            return True
+
+    def get_parameters(self):
+        parameters = []
+        for i in range(len(self.parents)):
+            if(isinstance(self.parents[i],ProbabilisticModel)):
+                for j in range(len(self.parents[i].value)):
+                    parameters.append(self.parents[i].value[j])
+        return parameters
+
+
+    @abstractmethod
+    def check_parameters(self, parameters):
+        raise NotImplementedError
+
+    @abstractmethod
+    def check_parameters_fixed(self, parameters):
+        raise NotImplementedError
+
+    @abstractmethod
+    def sample_from_distribution(self, k, rng=np.random.RandomState()):
```

Comment:
Here you want to forward simulate or sample the model, correct? Why not call the method just sample()